File: src/fly/anchors.ts
import type { Plan } from '../types'
import type { Anchor } from './ground'
import { resample } from './geo'

/* Everything in a Plan that needs a place on the real ground, as real lon/lat.
 * The flight and the headset both put the same things on the same tiles, so
 * they share this list and its keys. */

export const SAMPLE_STEP_M = 30
export const keyStop = (i: number) => `s${i}`
export const keyTarget = (i: number, id: string) => `t${i}:${id}`
export const keyLeg = (i: number, j: number) => `l${i}:${j}`

export function anchorsFor(plan: Plan): Anchor[] {
  const out: Anchor[] = [{ key: 'origin', ...plan.origin }]
  plan.stops.forEach((s, i) => {
    out.push({ key: keyStop(i), lat: s.lat, lon: s.lon })
    s.targets.forEach(t => out.push({ key: keyTarget(i, t.id), lat: t.lat, lon: t.lon }))
  })
  plan.legs.forEach((leg, i) => resample(leg.polyline, SAMPLE_STEP_M).forEach((p, j) => out.push({ key: keyLeg(i, j), ...p })))
  return out
}

File: quest/Assets/Orion/Runtime/Flight/Anchors.cs
using System.Collections.Generic;

namespace Orion.Flight
{
    /// <summary>Everything in a Day that needs a place on the real ground, as real lon/lat.</summary>
    public readonly struct Anchor
    {
        public const float SampleStepM = 30;

        public readonly string Key;
        public readonly double Lat, Lon;
        Anchor(string key, double lat, double lon) { Key = key; Lat = lat; Lon = lon; }

        public const string Origin = "origin";
        public static string Stop(int i) => $"s{i}";
        public static string Target(int i, string id) => $"t{i}:{id}";
        public static string Leg(int i, int j) => $"l{i}:{j}";

        public static List<Anchor> For(Day day)
        {
            var o = new List<Anchor> { new Anchor(Origin, day.origin.lat, day.origin.lon) };
            for (int i = 0; i < day.stops.Length; i++)
            {
                var s = day.stops[i];
                o.Add(new Anchor(Stop(i), s.lat, s.lon));
                foreach (var t in s.targets) o.Add(new Anchor(Target(i, t.id), t.lat, t.lon));
            }
            for (int i = 0; i < day.legs.Length; i++)
            {
                var pts = Geo.Resample(day.legs[i].polyline, SampleStepM);
                for (int j = 0; j < pts.Count; j++) o.Add(new Anchor(Leg(i, j), pts[j].lat, pts[j].lon));
            }
            return o;
        }
    }
}
